include/aarch32_gcc_accessor_macros.h

```diff
@@ -0,0 +1,243 @@
+// ----------------------------------------------------------------------------
+// This file contains aarch32-gcc specific implementations for reading and
+// writing system registers via all access mechanisms supported in AArch32
+// state. An "implementation" is the body of a function that carries out
+// a read/write using a specific access mechanism, used to back generated
+// accessor functions from a Shoulder generator
+
+#ifndef SHOULDER_AARCH32_GCC_ACCESSORS_H
+#define SHOULDER_AARCH32_GCC_ACCESSORS_H
+
+// Default implementation for reading a value using an instruction encoding.
+// Assumes that the instruction encoding contains r0 as the destination register
+// for the read
+//
+// @arg instruction_encoding a 32-bit binary representation of an instruction
+//      that reads data into r0 in aarch32 state
+//
+// returns the value read through the instruction
+#ifndef SHOULDER_AARCH32_ENCODED_READ_IMPL
+#define SHOULDER_AARCH32_ENCODED_READ_IMPL(instruction_encoding)               \
+    uint32_t val;                                                              \
+    __asm__ __volatile__(                                                      \
+        ".word "#instruction_encoding"\n"                                      \
+        "mov %[v], r0\n"                                                       \
+        : [v] "=r"(val)                                                        \
+        :                                                                      \
+        : "r0"                                                                 \
+    );                                                                         \
+    return val;
+#endif
+
+// Default implementation for writing a value using an instruction encoding.
+// Assumes that the instruction encoding contains r0 as the source register
+// for the write
+//
+// @arg instruction_encoding a 32-bit binary representation of an instruction
+//      that writes data from r0 in aarch32 state
+// @arg val an integer value to be written
+#ifndef SHOULDER_AARCH32_ENCODED_WRITE_IMPL
+#define SHOULDER_AARCH32_ENCODED_WRITE_IMPL(instruction_encoding, val)         \
+    __asm__ __volatile__(                                                      \
+        "mov r0, %[v]\n"                                                       \
+        ".word "#instruction_encoding"\n"                                      \
+        :                                                                      \
+        : [v] "r"(val)                                                         \
+        : "r0"                                                                 \
+    );
+#endif
+
+// Default implementation for reading a value using the LDR instruction.
+//
+// @arg addr the address to read from
+#ifndef SHOULDER_AARCH32_LDR_IMPL
+#define SHOULDER_AARCH32_LDR_IMPL(addr)                                        \
+    uint32_t address = addr;                                                   \
```

Comment:
Why is address redefined here? Does it insure that you are writing a 32bit integer and not a 64bit one?

Reply:
Exactly. Although```uint64_t``` is a type under 32-bit arm-gcc compilers, it uses funny tricks to split the value across multiple registers or the stack.